File: python-sdk/cleanster/webhook_utils.py
"""Utility functions for verifying Cleanster webhook delivery signatures.

When Cleanster sends a webhook, it computes an HMAC-SHA256 of the raw request
body using the ``secret`` returned when the webhook was created, and includes
the hex-encoded digest in the ``X-Webhook-Signature`` header.

Example::

    from cleanster import verify_webhook_signature

    @app.route("/webhook", methods=["POST"])
    def handle_webhook():
        secret    = os.environ["CLEANSTER_WEBHOOK_SECRET"]
        payload   = request.get_data(as_text=True)
        signature = request.headers.get("X-Webhook-Signature", "")

        if not verify_webhook_signature(secret, payload, signature):
            abort(401)
        # process event...
        return "", 200
"""

import hashlib
import hmac


def compute_webhook_signature(secret: str, payload: str) -> str:
    """Compute the HMAC-SHA256 of *payload* keyed with *secret*.

    Args:
        secret:  Webhook secret returned by the Cleanster API on creation.
        payload: Raw request body as a UTF-8 string.

    Returns:
        Lowercase hex-encoded HMAC-SHA256 digest.
    """
    return hmac.new(
        secret.encode("utf-8"),
        payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def verify_webhook_signature(secret: str, payload: str, signature: str) -> bool:
    """Verify that *signature* matches the expected HMAC-SHA256 of *payload*.

    Uses :func:`hmac.compare_digest` for constant-time comparison to prevent
    timing attacks.

    Args:
        secret:    Webhook secret returned by the Cleanster API.
        payload:   Raw request body as a UTF-8 string.
        signature: Value from the ``X-Webhook-Signature`` header.

    Returns:
        ``True`` if the signature is valid, ``False`` otherwise.
    """
    if not secret or not payload or not signature:
        return False
    expected = compute_webhook_signature(secret, payload)
    return hmac.compare_digest(expected, signature)

File: typescript-sdk/src/webhook-utils.ts
/**
 * Utility functions for verifying Cleanster webhook delivery signatures.
 *
 * When Cleanster sends a webhook, it computes an HMAC-SHA256 of the raw request
 * body using the `secret` returned when the webhook was created, and includes
 * the hex-encoded digest in the `X-Webhook-Signature` header.
 *
 * @example
 * ```typescript
 * import { verifyWebhookSignature } from "cleanster";
 *
 * // Express handler
 * app.post("/webhook", express.raw({ type: "application/json" }), (req, res) => {
 *   const secret    = process.env.CLEANSTER_WEBHOOK_SECRET!;
 *   const payload   = req.body.toString("utf8");
 *   const signature = req.headers["x-webhook-signature"] as string;
 *
 *   if (!verifyWebhookSignature(secret, payload, signature)) {
 *     return res.status(401).json({ error: "Invalid signature" });
 *   }
 *   // process event...
 *   res.sendStatus(200);
 * });
 * ```
 */

import { createHmac, timingSafeEqual } from "crypto";

/**
 * Compute the HMAC-SHA256 of `payload` keyed with `secret`.
 *
 * @param secret  Webhook secret returned by the Cleanster API on creation.
 * @param payload Raw request body as a UTF-8 string.
 * @returns Lowercase hex-encoded HMAC-SHA256 digest.
 */
export function computeWebhookSignature(secret: string, payload: string): string {
  return createHmac("sha256", secret).update(payload, "utf8").digest("hex");
}

/**
 * Verify that `signature` matches the expected HMAC-SHA256 of `payload`.
 * Uses a constant-time comparison to prevent timing attacks.
 *
 * @param secret    Webhook secret returned by the Cleanster API.
 * @param payload   Raw request body as a UTF-8 string.
 * @param signature Value from the `X-Webhook-Signature` header.
 * @returns `true` if the signature is valid, `false` otherwise.
 */
export function verifyWebhookSignature(secret: string, payload: string, signature: string): boolean {
  if (!secret || !payload || !signature) return false;
  const expected = Buffer.from(computeWebhookSignature(secret, payload), "utf8");
  const actual   = Buffer.from(signature, "utf8");
  if (expected.length !== actual.length) return false;
  return timingSafeEqual(expected, actual);
}
